### PR TITLE
Property with empty list 4

### PR DESCRIPTION
**Pull request by lucemia**
for issue #403

Previous PR
https://github.com/GoogleCloudPlatform/gcloud-python/pull/404

### CODE REVIEW

File: gcloud/datastore/connection.py [hunk 1]
```diff
@@ -429,6 +429,10 @@ def save_entity(self, dataset_id, key_pb, properties,
            will be replaced by those passed in ``properties``;  properties
            not passed in ``properties`` no longer be set for the entity.
 
+        .. note::
+           When saving an entity to the backend, property value set as
```
Comment (dhermes) [inline gcloud/datastore/connection.py]: Change to:
When saving an entity to the backend, a property value set as
an empty list or None cannot be saved and will be ignored.

File: gcloud/datastore/helpers.py [hunk 2]
```diff
@@ -218,6 +218,23 @@ def _get_value_from_property_pb(property_pb):
     return _get_value_from_value_pb(property_pb.value)
 
 
+def _set_protobuf_property(property_pb, name, value, indexed):
+    if isinstance(value, list) and len(value) == 0:
```
Comment (dhermes) [inline gcloud/datastore/helpers.py]: Where is `None` handled here?

File: gcloud/datastore/helpers.py [hunk 3]
```diff
@@ -218,6 +218,23 @@ def _get_value_from_property_pb(property_pb):
     return _get_value_from_value_pb(property_pb.value)
 
 
+def _set_protobuf_property(property_pb, name, value, indexed):
```
Comment (dhermes) [inline gcloud/datastore/helpers.py]: Add a docstring.

**Comment by dhermes** [comment]
1. I like this change. 
1. What is the difficulty in getting this to work for `PropertyFilter`?
1. Please run `tox` locally before submitting changes.
1. I want to let @tseaver have a chance to see this before signing off.

**Comment by tseaver** [comment]
`_set_protobuf_property` looks like a win:  I think we have a least a couple of versions of it lying around in the code.  Once @dhermes' issues are addressed, LGTM.

**Comment by lucemia** [comment]
@dhermes 
for `Property` with `None` value, my testing shows that the current behavior is the same as ndb. (both save a property will null value)
In that case we don't need to handle the `None` value.
I will updated the docstring for save_entity

# ndb
```
from google.appengine.ext import ndb

class Kind(ndb.Model):
    foo = ndb.StringProperty()
    bar = ndb.StringProperty()
    bar1 = ndb.StringProperty(repeated=True)


a = Kind(bar=None, bar1=[])
a.put()
```

```
from gcloud import datastore
from gcloud.credentials import *
from gcloud.datastore.key import *


client_email = "1065106153444-mdj042q86ciof489e71i49joe6an21k6@developer.gserviceaccount.com"
private_key_path = "tagtooadex2-9e2928ac0acf.p12"
SCOPE = ('https://www.googleapis.com/auth/datastore ',
         'https://www.googleapis.com/auth/userinfo.email')
credentials = get_for_service_account_p12(client_email, private_key_path, SCOPE)

key_pb = Key('Kind', 1234, dataset_id="tagtooadex2").to_protobuf()

conn = datastore.connection.Connection(credentials)

result = conn.save_entity("tagtooadex2", key_pb, {
    'foo': u'Foo',
    'bar': None,
    'bar1': []
})
```

**Comment by dhermes** [comment]
@lucemia I was curious to see what actually is stored, which requires inspecting the retrieved protobuf.

After storing
```python
>>> from gcloud import datastore
>>> from gcloud.datastore.entity import Entity
>>> from gcloud.datastore.key import Key
>>>
>>> datastore.set_default_connection()
>>> dataset_id = 'datasetfoo'
>>> datastore.set_default_dataset_id(dataset_id)
>>>
>>> key = Key('Kind', 1234)
>>> entity = Entity(key=key)
>>> entity['foo'] = None
>>> dict(entity)
{'foo': None}
>>> entity.save()
```

retrieving the same entity shows that the property is set but `Property.value` is empty:

```python
>>> entity_pb, = datastore.get_connection().lookup(
...     dataset_id=dataset_id,
...     key_pbs=[key.to_protobuf()],
... )
>>> prop_foo, = entity_pb.property
>>> prop_foo.name
u'foo'
>>> prop_foo_value = prop_foo.value
>>> prop_foo_value._fields
{}
```

----

This happens because in [`helpers._set_protobuf_value`][1], `Property.value` is cleared but the property is still set on the `Entity`.

It's unclear what is the right thing to support, maybe going with `ndb` is fine? However, the note you introduce makes it seem the property will be dropped all together (as is done for `[]`).

----

Cleanup:

```python
>>> print entity_pb
key {
  partition_id {
    dataset_id: "s~datasetfoo"
  }
  path_element {
    kind: "Kind"
    id: 1234
  }
}
property {
  name: "foo"
  value {
  }
}

>>> entity.key.delete()
```

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/0525530ad7a76e408ca2e842b883d49c353ed56d/gcloud/datastore/helpers.py#L238

**Comment by lucemia** [comment]
@dhermes 
> It's unclear what is the right thing to support, maybe going with ndb is fine? However, the note you introduce makes it seem the property will be dropped all together (as is done for []).

Currently I go with ndb way. I updated the note to only include empty list case.

> What is the difficulty in getting this to work for PropertyFilter?

Store a `Property` whose value is empty list sounds normal and should not raise any exception (currently I am working with).
But create a `PropertyFilter` whose value is empty list sounds weird. 
Which behavior (write a helper function or raise exception) is recommend for PropertyFilter case is unclear for me. 

ps. Rebase is welcome

**Comment by dhermes** [comment]
I see, I was thinking the issue was the code, but your saying the expected behavior is a bit subtle.

Also, what do you mean by "Rebase is welcome"?

**Comment by lucemia** [comment]
@dhermes 
> Also, what do you mean by "Rebase is welcome"?

[Last pull request](https://github.com/GoogleCloudPlatform/gcloud-python/pull/404#issuecomment-68409980) you ask about re-writeing history. I mean feel free to do so. :)

**Comment by lucemia** [comment]
I have squash the commit

**Pull request merged**